packages/calico/extra/calico-cni-wrapper.go

```diff
@@ -0,0 +1,188 @@
+	sigs := make(chan os.Signal, 1)
+	signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)
+	go func() {
+		for {
+			sig := <-sigs
+			if cmd.ProcessState != nil && cmd.ProcessState.Exited() {
+				return
+			}
+			cmd.Process.Signal(sig)
+		}
+	}()
+
+	// Send STDIN and release
+	io.WriteString(stdin, stdinStr)
+	stdin.Close()
+
+	// Wait until the command is completed and remove the signal handlers
+	err = cmd.Wait()
+	signal.Reset(syscall.SIGINT, syscall.SIGTERM)
+	sigs <- syscall.SIGINT
+
+	if err != nil {
+		// Get exit code on non-zero exits
+		if exiterr, ok := err.(*exec.ExitError); ok {
+			if status, ok := exiterr.Sys().(syscall.WaitStatus); ok {
+				return status.ExitStatus(), nil
+			}
+		} else {
+			return 0, err
+		}
+	}
+
+	return 0, nil
+}
+
+func parseStdinJSONMap() (map[string]interface{}, error) {
+	dst := make(map[string]interface{})
+
+	buf := new(bytes.Buffer)
+	buf.ReadFrom(os.Stdin)
+
+	err := json.Unmarshal(buf.Bytes(), &dst)
+	if err != nil {
+		return nil, err
+	}
+
+	return dst, nil
+}
+
+func compressLongLabel(str string) string {
+	h := sha256.New()
+	h.Write([]byte(str))
+
+	hash := fmt.Sprintf("%x", h.Sum(nil))
+	lastPart := str[len(str)-53:]
+
+	return fmt.Sprintf("%s...%s", hash[0:7], lastPart)
+}
+
+func compressLabels(labels []map[string]interface{}) []map[string]interface{} {
+	for _, kv := range labels {
+		if value, ok := kv["value"]; ok {
+			if valueStr, ok := value.(string); ok {
+				if len(valueStr) > 63 {
+					kv["value"] = compressLongLabel(valueStr)
+				}
+			}
+		}
+	}
+	return labels
+}
+
+func compressMesosArgsConfig(config map[string]interface{}) map[string]interface{} {
+	//
+	// We are manually parsing the mesos args interface, as defined here:
+	// http://mesos.apache.org/documentation/latest/cni/#passing-network-labels-and-port-mapping-information-to-cni-plugi
+
+	// .args{}
+	if argsMapIface, ok := config["args"]; ok {
+		if argsMap, ok := argsMapIface.(map[string]interface{}); ok {
+			// .args."org.apache.mesos"{}
+			if mesosMapIface, ok := argsMap["org.apache.mesos"]; ok {
+				if mesosMap, ok := mesosMapIface.(map[string]interface{}); ok {
+					// .args."org.apache.mesos".network_info{}
+					if netInfoMapIface, ok := mesosMap["network_info"]; ok {
+						if netInfoMap, ok := netInfoMapIface.(map[string]interface{}); ok {
+							// .args."org.apache.mesos".network_info.labels{}
+							if labelsMapIface, ok := netInfoMap["labels"]; ok {
+								if labelsMap, ok := labelsMapIface.(map[string]interface{}); ok {
+									// .args."org.apache.mesos".network_info.labels.labels[]
+									if labelsSliceIface, ok := labelsMap["labels"]; ok {
+										if labelsSlice, ok := labelsSliceIface.([]interface{}); ok {
+
+											// Convert interface to map labels and compress
+											var labelMapArray []map[string]interface{} = nil
+											for _, val := range labelsSlice {
+												if v, ok := val.(map[string]interface{}); ok {
```

Comment:
![](https://pbs.twimg.com/media/D5tgy4cWsAMk3kf?format=png&name=small)

Reply:
https://github.com/tidwall/gjson#get-a-value

Reply:
hehehe, I was trying to avoid any external dependencies since we are not using complete go build environment, but what the heck 😄

Reply:
Nope, `gjson` won't work. (1) I am traversing this map to _Update_ not read, and (2) I am operating on an already parsed JSON object (map of interfaces), not string.

Reply:
I agree that the "more beautiful" solution here would be to use a map traversal utility, but (I think that)the fact that this is only a temporary fix till the next release does not really justify the effort.